### PR TITLE
Add tests for ID and Type colisions

### PR DESCRIPTION
**Pull request by mapedraza**

### CODE REVIEW

**Comment by AlvaroVega** [comment]
With this branch the fix of https://github.com/telefonicaid/iotagent-node-lib/pull/1483 I'm getting some errors:

  707 passing (41s)
  11 pending
  3 failing

  1) NGSI-v2 - Active attributes test
       When the IoT Agent receives new information from a device and CBis defined using environment variables
         should change the value of the corresponding attribute in the context broker:
     Error: Timeout of 8000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/avega/tme/fiware/iotagent-node-lib/test/unit/ngsiv2/ngsiService/active-devices-test.js)
      at listOnTimeout (node:internal/timers:564:17)
      at process.processTimers (node:internal/timers:507:7)

  2) NGSI-v2 - Active attributes test
       When the IoT Agent receives autoprovisioned id and type measures
         should not affect to the real ID and Type to store in the context broker:
     Uncaught AssertionError: expected RequestError {
  name: 'RequestError',
  code: 'ENOTFOUND',
  timings: Object {
    start: 1696262045323,
    socket: 1696262045323,
    lookup: 1696262045325,
    connect: undefined,
    secureConnect: undefined,
    upload: undefined,
    response: undefined,
    end: undefined,
    error: 1696262045325,
    abort: undefined,
    phases: Object {
      wait: 0,
      dns: 2,
      tcp: undefined,
      tls: undefined,
      request: undefined,
      firstByte: undefined,
      download: undefined,
      total: 2
    }
  },
  message: 'getaddrinfo ENOTFOUND cbhost'
} to not exist
      at /home/avega/tme/fiware/iotagent-node-lib/test/unit/ngsiv2/ngsiService/active-devices-test.js:958:28
      at /home/avega/tme/fiware/iotagent-node-lib/lib/services/ngsi/entities-NGSI-v2.js:86:497
      at /home/avega/tme/fiware/iotagent-node-lib/lib/request-shim.js:64:728
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

  3) NGSI-v2 - Active attributes test
       When the IoT Agent receives provisioned id and type measures
         should not affect to the real ID and Type to store in the context broker:
     Error: Timeout of 8000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/avega/tme/fiware/iotagent-node-lib/test/unit/ngsiv2/ngsiService/active-devices-test.js)
      at listOnTimeout (node:internal/timers:564:17)
      at process.processTimers (node:internal/timers:507:7)

**Comment by mapedraza** [comment]
3 Cases:

**Case 1** Receiving measures (`id` and `type`) not provisioned (autoprovision)

In this case, overwrites the `id` and `type` values with the measured received. `id` and `type` are converted to JSON objects

```json
{
        "id": {
            "value": "idIoTA",
            "type": "text"
        },
        "type": {
            "value": "typeIoTA",
            "type": "text"
        },
        "meas": {
            "value": "measIoTA",
            "type": "String"
        }
    }

```

**Case 2**. Mapping measures (`i` and `t`) to `id` and `type` attributes

In this case, it ignores the mappings, working as expected

**Case 3**. `object_id` = `id` and `type`

In this case, it ignores `id` and `type`, and they are not present in the request.

```json
[{
    "url": "http://192.168.1.1:1026/v2/op/update",
    "method": "POST",
    "headers": {
        "fiware-service": "smartgondor",
        "fiware-servicepath": "gardens"
    },
    "json": {
        "entities": [
            {
                "meas": {
                    "value": "measIoTA",
                    "type": "String"
                }
            }
        ],
        "actionType": "append"
    }
}]
```

**Comment by mapedraza** [comment]
I suggest to change this PR destination to branch `fix/remove_measures_with_attribute_name_id_type` (PR #1483)

**Comment by mapedraza** [comment]
> 7197f62

After https://github.com/telefonicaid/iotagent-node-lib/pull/1482/commits/7197f627e5d522e046d249a005cc83ed1b137c05

Only 1 test is failing with your branch code

**Comment by AlvaroVega** [comment]
Invalid mapping is detected in case 3:

time=2023-10-03T06:22:18.522Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 invalid mapping for attr={"name":"id","object_id":"id","type":"text"}  newAttr="stupiddevice3"

**Comment by mapedraza** [comment]
> Invalid mapping is detected in case 3:
> 
> time=2023-10-03T06:22:18.522Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 invalid mapping for attr={"name":"id","object_id":"id","type":"text"} newAttr="stupiddevice3"

The following log is printed. A request without ID and type is executed:

```
time=2023-10-03T06:34:56.023Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 currentEntity sorted {"id":"stupiddevice3","type":"StupidDevice2","meas":{"value":"measIoTA","type":"text"}} 
time=2023-10-03T06:34:56.023Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 invalid mapping for attr={"name":"id","object_id":"id","type":"text"}  newAttr="stupiddevice3"
time=2023-10-03T06:34:56.023Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 procesing j="id" attr=undefined ctxt={"meas":"measIoTA","type":"StupidDevice2"} newAttr=undefined 
time=2023-10-03T06:34:56.024Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 after procesing attr="id"  current entity={"type":"StupidDevice2","meas":{"value":"measIoTA","type":"text"}}  current payload={"entities":[{"type":"StupidDevice2","meas":{"value":"measIoTA","type":"text"}}],"actionType":"append"}
time=2023-10-03T06:34:56.024Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 invalid mapping for attr={"name":"type","object_id":"type","type":"text"}  newAttr="StupidDevice2"
time=2023-10-03T06:34:56.024Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 procesing j="type" attr=undefined ctxt={"meas":"measIoTA","type":"StupidDevice2"} newAttr=undefined 
time=2023-10-03T06:34:56.024Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 after procesing attr="type"  current entity={"meas":{"value":"measIoTA","type":"text"}}  current payload={"entities":[{"meas":{"value":"measIoTA","type":"text"}}],"actionType":"append"}
time=2023-10-03T06:34:56.024Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 procesing j="meas" attr={"name":"meas","object_id":"meas","type":"String"} ctxt={"meas":"measIoTA","type":"StupidDevice2"} newAttr={"value":"measIoTA","type":"text"} 
time=2023-10-03T06:34:56.024Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 after procesing attr="meas"  current entity={"meas":{"value":"measIoTA","type":"String"}}  current payload={"entities":[{"meas":{"value":"measIoTA","type":"String"}}],"actionType":"append"}
time=2023-10-03T06:34:56.024Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 ending payload={"entities":[{"meas":{"value":"measIoTA","type":"String"}}],"actionType":"append"}
time=2023-10-03T06:34:56.024Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 payload with native types and without object_id={"entities":[{"meas":{"value":"measIoTA","type":"String"}}],"actionType":"append"}
time=2023-10-03T06:34:56.024Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=Not updating device value in the Context Broker at [http://192.168.1.1:1026/v2/op/update] due to empty payload 

[{
    "url": "http://192.168.1.1:1026/v2/op/update",
    "method": "POST",
    "headers": {
        "fiware-service": "smartgondor",
        "fiware-servicepath": "gardens"
    },
    "json": {
        "entities": [
            {
                "meas": {
                    "value": "measIoTA",
                    "type": "String"
                }
            }
        ],
        "actionType": "append"
    }
}]
```

File: test/unit/ngsiv2/ngsiService/active-devices-test.js [hunk 1]
```diff
@@ -856,6 +902,153 @@ describe('NGSI-v2 - Active attributes test', function () {
...
+            },
+            {
+                name: 'm',
+                type: 'text',
+                value: 'measIoTA'
+            }
+        ];
+
+        beforeEach(function (done) {
+            // logger.setLevel('DEBUG');
```
Comment (fgalan) [inline test/unit/ngsiv2/ngsiService/active-devices-test.js:925]:
```suggestion
```

Lefover?

File: test/unit/ngsiv2/ngsiService/active-devices-test.js [hunk 2]
```diff
@@ -856,6 +902,153 @@ describe('NGSI-v2 - Active attributes test', function () {
...
+            },
+            {
+                name: 'm',
+                type: 'text',
+                value: 'measIoTA2'
+            }
+        ];
+
+        beforeEach(function (done) {
+            //logger.setLevel('DEBUG');
```
Comment (fgalan) [inline test/unit/ngsiv2/ngsiService/active-devices-test.js:974]:
```suggestion
```

Leftover?

File: test/unit/ngsiv2/ngsiService/active-devices-test.js [hunk 3]
```diff
@@ -856,6 +902,153 @@ describe('NGSI-v2 - Active attributes test', function () {
...
+            },
+            {
+                name: 'meas',
+                type: 'text',
+                value: 'measIoTA'
+            }
+        ];
+
+        beforeEach(function (done) {
+            // logger.setLevel('DEBUG');
```
Comment (fgalan) [inline test/unit/ngsiv2/ngsiService/active-devices-test.js:1023]:
```suggestion
```

Leftover?

**Comment by AlvaroVega** [comment]
LGTM

**Comment by AlvaroVega** [comment]
> > Invalid mapping is detected in case 3:
> > time=2023-10-03T06:22:18.522Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 invalid mapping for attr={"name":"id","object_id":"id","type":"text"} newAttr="stupiddevice3"
> 
> The following log is printed at https://github.com/telefonicaid/iotagent-node-lib/tree/fix/remove_measures_with_attribute_name_id_type / [cadda42](https://github.com/telefonicaid/iotagent-node-lib/commit/cadda42d8e9ac1e1b5eb74aa8337640267dfbc88) . A request without ID and type is executed:
> 
> ```
> time=2023-10-03T06:34:56.023Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 currentEntity sorted {"id":"stupiddevice3","type":"StupidDevice2","meas":{"value":"measIoTA","type":"text"}} 
> time=2023-10-03T06:34:56.023Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 invalid mapping for attr={"name":"id","object_id":"id","type":"text"}  newAttr="stupiddevice3"
> time=2023-10-03T06:34:56.023Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 procesing j="id" attr=undefined ctxt={"meas":"measIoTA","type":"StupidDevice2"} newAttr=undefined 
> time=2023-10-03T06:34:56.024Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 after procesing attr="id"  current entity={"type":"StupidDevice2","meas":{"value":"measIoTA","type":"text"}}  current payload={"entities":[{"type":"StupidDevice2","meas":{"value":"measIoTA","type":"text"}}],"actionType":"append"}
> time=2023-10-03T06:34:56.024Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 invalid mapping for attr={"name":"type","object_id":"type","type":"text"}  newAttr="StupidDevice2"
> time=2023-10-03T06:34:56.024Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 procesing j="type" attr=undefined ctxt={"meas":"measIoTA","type":"StupidDevice2"} newAttr=undefined 
> time=2023-10-03T06:34:56.024Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 after procesing attr="type"  current entity={"meas":{"value":"measIoTA","type":"text"}}  current payload={"entities":[{"meas":{"value":"measIoTA","type":"text"}}],"actionType":"append"}
> time=2023-10-03T06:34:56.024Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 procesing j="meas" attr={"name":"meas","object_id":"meas","type":"String"} ctxt={"meas":"measIoTA","type":"StupidDevice2"} newAttr={"value":"measIoTA","type":"text"} 
> time=2023-10-03T06:34:56.024Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 after procesing attr="meas"  current entity={"meas":{"value":"measIoTA","type":"String"}}  current payload={"entities":[{"meas":{"value":"measIoTA","type":"String"}}],"actionType":"append"}
> time=2023-10-03T06:34:56.024Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 ending payload={"entities":[{"meas":{"value":"measIoTA","type":"String"}}],"actionType":"append"}
> time=2023-10-03T06:34:56.024Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=sendUpdateValueNgsi2 payload with native types and without object_id={"entities":[{"meas":{"value":"measIoTA","type":"String"}}],"actionType":"append"}
> time=2023-10-03T06:34:56.024Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.Entities-v2 | msg=Not updating device value in the Context Broker at [http://192.168.1.1:1026/v2/op/update] due to empty payload 
> 
> [{
>     "url": "http://192.168.1.1:1026/v2/op/update",
>     "method": "POST",
>     "headers": {
>         "fiware-service": "smartgondor",
>         "fiware-servicepath": "gardens"
>     },
>     "json": {
>         "entities": [
>             {
>                 "meas": {
>                     "value": "measIoTA",
>                     "type": "String"
>                 }
>             }
>         ],
>         "actionType": "append"
>     }
> }]
> ```

fixed in https://github.com/telefonicaid/iotagent-node-lib/pull/1483/commits/cadda42d8e9ac1e1b5eb74aa8337640267dfbc88

**Pull request merged**